proposals/0002.md

```diff
@@ -0,0 +1,318 @@
+* Title: Structured error handling in Rust
+* Champion: @CodeSandwich
+* Conversation (optional): [thread on radicle.community](https://radicle.community/t/proposal-structured-error-handling-in-rust/179)
+* References:
+  * [The previous error handling proposal](https://github.com/radicle-dev/radicle-decisions/blob/master/proposals/0001.md)
+  * [`std::error::Error`](https://doc.rust-lang.org/std/error/trait.Error.html)
+  * [A discussion about the good design of errors](https://users.rust-lang.org/t/do-people-not-care-about-printable-error-chains-a-k-a-how-to-nicely-implement-display-for-an-error/35362)
+
+# Summary/Motivation/Introduction
+We need good error handling to make our lives easier.
+We've already agreed to use `thiserror` and `anyhow` in the
+[previous error handling proposal](https://github.com/radicle-dev/radicle-decisions/blob/master/proposals/0001.md).
+They are fantastic tools, but we're still lacking a common design language.
+There are many ways to create an error and some of them don't play well with others.
+Without common conventions handling errors is problematic too.
+
+This is a guide to structuring errors well.
+
+# Explanation
+
+## The great `Error` unification
+
+The ultimate goal is to make your errors as easy to integrate and handle as possible.
+The canonical way of achieving this is by implementing the `std::error::Error` trait.
+It's universally used in the Rust ecosystem and provides a common interface for working with
+different errors coming from numerous crates.
+It's not possible to enforce the exact behavior of an implementation of the `Error` trait, but
+there's a widely used convention making it reasonably uniform and predictable.
+
+## The correct behavior of an `Error` implementation
+
+`Error` provides 2 core functionalities:
+
+- it provides the `source` method, which may or may not return a reference to another
+`std::error::Error`
+- it derives from `Display` making all its implementors printable
+
+### The behavior of `source`
+
+`source` should return a value when the error has another error nested inside.
+The wrapped error should be the cause of the outer one, usually a bit lower level and more detailed.
+The errors can be wrapped multiple times to express the whole chain of errors down to the root
+cause. It can be traversed by calling `source` recursively.
+
+For example the custom `FileLoadError` can wrap the `std::io::Error`.
+This way the file loading function doesn't have to know about all the possible file system failures
+described by the `std::io::Error`.
+When it gets one, it simply wraps it in the `FileLoadError` and returns it.
+If a higher-level function is interested in what caused the `FileLoadError`, it can call its `source`
+method and get its cause.
+
+`source` should return `None` when the error is the root cause and doesn't wrap another error.
+It's also impossible to return a wrapped error when it doesn't implement `Error`.
```

Comment:
```suggestion
All wrapped errors are also implementations of `Error`.
```

Reply:
But this is not true, the causality is reversed.